File: android/src/main/java/com/gaya/digital_turbine_plugin/BannerView.java
package com.gaya.digital_turbine_plugin;

import android.app.Activity;
import android.content.Context;
import android.view.View;
import android.widget.FrameLayout;
import android.view.ViewGroup;
import com.fyber.fairbid.ads.Banner;
import com.fyber.fairbid.ads.ImpressionData;
import com.fyber.fairbid.ads.banner.BannerError;
import com.fyber.fairbid.ads.banner.BannerListener;
import com.fyber.fairbid.ads.banner.BannerOptions;
import com.fyber.fairbid.ads.banner.BannerSize;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.platform.PlatformView;

import java.util.HashMap;
import java.util.Map;

public class BannerView implements PlatformView, MethodChannel.MethodCallHandler {

    private FrameLayout containerView;
    private String placementId;
    private MethodChannel methodChannel;
    private Activity activity;

    BannerView(Context context, Activity activity, int id, Map<String, Object> creationParams, BinaryMessenger messenger) {
        this.activity = activity;
        this.containerView = new FrameLayout(context);
        this.placementId = (String) creationParams.get("placementId");
        this.methodChannel = new MethodChannel(messenger, "digital_turbine_banner_view_" + id);
        this.methodChannel.setMethodCallHandler(this);
        setupBannerListener();
        loadAd();
    }

    @Override
    public View getView() {
        return containerView;
    }

    @Override
    public void dispose() {
        methodChannel.setMethodCallHandler(null);
        Banner.destroy(placementId);
    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        switch (call.method) {
            case "loadAd":
                loadAd();
                result.success(null);
                break;
            case "disposeAd":
                Banner.destroy(placementId);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void setupBannerListener() {
        Banner.setBannerListener(new BannerListener() {
            @Override
            public void onError(String placementId, BannerError error) {
                Map<String, Object> args = new HashMap<>();
                args.put("placementId", placementId);
                args.put("error", error.toString());
                methodChannel.invokeMethod("onBannerError", args);
            }

            @Override
            public void onLoad(String placementId) {
                methodChannel.invokeMethod("onBannerLoad", createArguments(placementId, null));
            }

            @Override
            public void onShow(String placementId, ImpressionData impressionData) {
                methodChannel.invokeMethod("onBannerShow", createArguments(placementId, impressionData));
            }

            @Override
            public void onClick(String placementId) {
                methodChannel.invokeMethod("onBannerClick", createArguments(placementId, null));
            }

            @Override
            public void onRequestStart(String placementId, String requestId) {
                Map<String, Object> args = new HashMap<>();
                args.put("placementId", placementId);
                args.put("requestId", requestId);
                methodChannel.invokeMethod("onBannerRequestStart", args);
            }
        });
    }

    private void loadAd() {
        if (activity != null) {
            BannerOptions options = new BannerOptions()
                    .withSize(BannerSize.MREC).placeInContainer(containerView);

            activity.runOnUiThread(() -> {
                Banner.show(placementId, options, activity);
            });
        } else {
            System.err.println("Activity is null, cannot load banner ad");
        }
    }

    private Map<String, Object> createArguments(String placementId, ImpressionData impressionData) {
        Map<String, Object> args = new HashMap<>();
        args.put("placementId", placementId);
        if (impressionData != null) {
            args.put("impressionData", impressionData.getJsonString());
        }
        return args;
    }
}

File: android/src/main/kotlin/com/gaya/digital_turbine_plugin/BannerView.kt
// BannerView.kt
package com.gaya.digital_turbine_plugin

import android.app.Activity
import android.content.Context
import android.view.View
import com.fyber.fairbid.ads.banner.BannerView as FairBidBannerView
import com.fyber.fairbid.ads.banner.BannerListener
import com.fyber.fairbid.ads.banner.BannerOptions
import com.fyber.fairbid.ads.ImpressionData
import com.fyber.fairbid.ads.banner.BannerError
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class BannerViewWrapper(
    context: Context,
    private val activity: Activity,
    viewId: Int,
    creationParams: Map<String, Any>?,
    messenger: BinaryMessenger
) : PlatformView, MethodChannel.MethodCallHandler {

    private val bannerView: FairBidBannerView
    private val methodChannel: MethodChannel
    private val placementId: String

    init {
        placementId = creationParams?.get("placementId") as? String ?: ""
        bannerView = FairBidBannerView(context, placementId)
        methodChannel = MethodChannel(messenger, "digital_turbine_banner_view_1")
        methodChannel.setMethodCallHandler(this)
        setupBannerListener()
    }

    override fun getView(): View = bannerView

    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
        bannerView.destroy()
    }

    override fun onMethodCall(call : MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadAd" -> {
                loadAd()
                result.success(null)
            }
            "disposeAd" -> {
                dispose()
                result.success("BANNER_DISPOSED")
            }
            else -> result.notImplemented()
        }
    }

    private fun setupBannerListener() {
        bannerView.bannerListener = object : BannerListener {
            override fun onError(placementId: String, error: BannerError) {
                val args = mapOf(
                    "placementId" to placementId,
                    "error" to error.toString()
                )
                methodChannel.invokeMethod("onBannerError", args)
            }

            override fun onLoad(placementId: String) {
                methodChannel.invokeMethod("onBannerLoad", mapOf("placementId" to placementId))
            }

            override fun onShow(placementId: String, impressionData: ImpressionData) {
                methodChannel.invokeMethod("onBannerShow", mapOf(
                    "placementId" to placementId,
                    "impressionData" to impressionData.jsonString
                ))
            }

            override fun onClick(placementId: String) {
                methodChannel.invokeMethod("onBannerClick", mapOf("placementId" to placementId))
            }

            override fun onRequestStart(placementId: String, requestId: String) {
                val args = mapOf(
                    "placementId" to placementId,
                    "requestId" to requestId
                )
                methodChannel.invokeMethod("onBannerRequestStart", args)
            }
        }
    }

    private fun loadAd() {
        val options = BannerOptions()
        bannerView.load(options)
    }
}
